celery/app/base.py

```diff
@@ -870,7 +870,8 @@ def prepare_config(self, c):
 
     def now(self):
         """Return the current time and date as a datetime."""
-        return self.loader.now(utc=self.conf.enable_utc)
+        from datetime import datetime
+        return datetime.utcnow().replace(tzinfo=self.timezone)
```

Comment:
I think this is wrong. This will not return the current time in `self.timezone`. Let's say for example that `self.timezone` is +8, at it is currently `12:00 UTC`. This method will return `12:00 +8`, which is the same point in time as `04:00 UTC` – not `12:00 UTC`.
Note that this current implementation might even return nonsensical times, like `2017-03-25 02:30` for a Europe/Berlin timezone – a time that doesn't exist.

The correct way to get the current time in a local timezone is `self.timezone.localize(datetime.utcnow().replace(tzinfo=pytz.utc))`. See the [same functionality in django](https://github.com/django/django/blob/master/django/utils/timezone.py#L191) and the [pytz docs](http://pytz.sourceforge.net/).

But you can't just make a function that used to return utc times return times from timezones that observe DST. If the caller code is not aware of that, things might break. Example: You return `2017-03-24 03:30 Europe/Berlin`. The caller adds 24h to this and then checks if the time is before `04:00`. Adding 24h crosses a DST boundary, so the above +24h is actually `2017-03-25 04:30`. But if you add 24h to a `2017-03-24 03:30 Europe/Berlin` datetime object, you actually get `2017-03-25 03:30 Europe/Berlin [NON-DST]`. You need to call `timezone.localize` on that to turn that into a sensible datetime object – see the [pytz docs](http://pytz.sourceforge.net/#localized-times-and-date-arithmetic).

Hope this helps.

Reply:
@Chronial would you mind sending a pr with your suggestions?

Reply:
Providing a correct implementation using timezones is not easy. Since I moved away from using celery (its broken timzone handling beeing one of the reasons ^^), I unfortunately do not have the time to provide a full PR for this.

Just wanted to report this bug, since I stumbled upon it while browsing.

Reply:
@Chronial you are right, this has already been corrected by other contributors in [master/latest release](https://github.com/celery/celery/blob/master/celery/app/base.py#L885).

Reply:
Ah, sorry for digging up this old PR. I came here via the mention on #4006 and didn't realize how old this PR was.

Reply:
No problem @Chronial thanks for the feedback.